Batch of signal-related updates

alxhub opened the pull request:
(more coming)

alxhub committed:
feat(core): add `assertInInjectionContext`

This commit adds an assertion function to the public API, which allows
authors of functions which rely on `inject` to validate that they're being
called with the right context. This mostly produces a nicer error message
than calling `inject()` and relying on Angular's default error message for
that.

alxhub committed:
refactor(core): rename `SettableSignal` to `WritableSignal`

This commit renames the `SettableSignal` interface to align with recent
naming changes. `WritableSignal` is considered a more suitable name since
it supports more mutation operations than just directly setting.

alxhub committed:
refactor(core): switch `signal` and `computed` to take options argument

This commit switches the `signal` and `computed` APIs to accept an optional
options argument as their second argument, instead of an equality function
directly. The equality function has moved to an option in the options
argument.

alxhub committed:
refactor(core): move `effect` to render3 and add `DestroyRef` integration

The `effect` implemented in the signal library is useful for testing but
does not integrate with Angular. This commit moves that code to the
actual framework package and integrates it with automatic cleanup via
`DestroyRef`. A simpler effect implementation is used in the signal tests to
test the `Watch` primitive.

Further commits will update the scheduling to tie effects together with
change detection.

alxhub committed:
refactor(core): make signal values deeply immutable at the type level

This commit adds a `DeepReadonly` type to the signals API, and makes signal
getters return an immutable version of their value type. This doesn't
prevent all mutation but adds some friction against modifying the values
within signals outside of the proper mutation APIs.

AndrewKushnir commented on packages/core/src/di/contextual.ts:
```diff
@@ -37,3 +38,22 @@ export function runInInjectionContext<ReturnT>(injector: Injector, fn: () => Ret
+ *
+ * @publicApi
+ */
+export function assertInInjectionContext(debugFn: Function): void {
```
Just as an idea: we can consider exposing a function like `isInInjectionContext`, which would return true/false. Using this function developer may be able to implement the necessary assertions and/or other logic as needed (and it's a bit easier to work with vs a function that throws).

AndrewKushnir commented on packages/core/src/render3/reactivity/effect.ts:
```diff
@@ -0,0 +1,116 @@
+    globalWatches.delete(watch);
+  };
+
+  unregisterOnDestroy = destroyRef?.onDestroy(destroy);
```
I'm a bit confused on why do we keep track of the unregister fn and invoke it on destroy. Is it possible that the same callback would be invoked multiple times and we want to make sure it's always invoked once?

alxhub committed:
refactor(core): run effects in the zone in which they're created

Previously, effects were queued in the global microtask queue and executed
directly. This had an undesired consequence: whichever effect scheduled
first determined the zone in which all other effects would run, *and* that
zone depended on where the signal happened to be set which triggered that
first effect. This behavior would be extremely unpredictable.

This commit adds zone awareness to the effect API. effects now capture the
current zone when they're created, and this zone is used to run the effect
callback regardless of which zone set the signal.

alxhub committed:
refactor(core): unify signal producer/consumer into nodes

Previously the signals implementation maintained separate interfaces for
`Producer` and `Consumer` nodes, with implementers choosing to implement
one or both interfaces. Operations defined against those interfaces were
exposed as importable functions to be called with the object implementing
the relevant contract as the first argument.

This commit refactors the implementation to merge both abstractions into a
single `ReactiveNode` base class, which represents both producers and
consumers in the graph. Implementers choose to interact with a subset of the
`ReactiveNode` API depending on their role in the graph. Operations are now
available as protected methods on the base class, instead of separate
functions.

alxhub committed:
refactor(core): expose `ReactiveNode` privately on signal functions

Signal functions are marked with a `[SIGNAL]` symbol property, which was
previously set to `true`.

This commit changes the `[SIGNAL]` property to have the `ReactiveNode` as
its value instead. This doesn't change the behavior of the signal library
itself, but exposes the signal internals in a way where developer tooling
can look at any signal and interact with its dependency graph.

pkozlowski-opensource commented on packages/core/src/signals/src/signal.ts:
```diff
@@ -96,13 +96,13 @@ class SettableSignalImpl<T> implements Producer {
                      update: signalNode.update.bind(signalNode),
                      mutate: signalNode.mutate.bind(signalNode),
-                   }) as unknown as SettableSignal<T>;
+                   }) as unknown as WritableSignal<T>;
```
super-nit: we've got the `// Casting here is required for g3` which is pretty enigmatic - could we expand it to say "why" it is needed in g3?

alxhub replied:
Done!

pkozlowski-opensource commented on packages/core/src/signals/src/signal.ts:
```diff
@@ -96,8 +96,8 @@ class WritableSignalImpl<T> implements Producer {
 export function signal<T>(
-    initialValue: T, equal: ValueEqualityFn<T> = defaultEquals): WritableSignal<T> {
-  const signalNode = new WritableSignalImpl(initialValue, equal);
+    initialValue: T, options?: {equal?: ValueEqualityFn<T>}): WritableSignal<T> {
```
nit / thinking out loud: should we introduce a specific type for options? Or is it just too much hassle / too many symbols in the public API surface?

alxhub replied:
Done :)

pkozlowski-opensource reviewed (commented):
@alxhub I went through this PR and most changes LGTM. Is there anything that you think is still needed to get this PR over the finish line?

One thing I'm seeing as still needing work is the G3 failures that seem legit (hence adding cleanup label). Also there are open comments from @AndrewKushnir .

pkozlowski-opensource commented on packages/core/src/render3/reactivity/effect.ts:
```diff
@@ -0,0 +1,116 @@
+ *
+ * @developerPreview
+ */
+export function effect(effectFn: () => void, options?: CreateEffectOptions): EffectRef {
```
change required: since you are moving the effect's implementation here, it should be deleted from the previous location.

alxhub replied:
Done!

alxhub commented on packages/core/src/render3/instructions/shared.ts:
```diff
@@ -1806,6 +1818,9 @@ export function detectChangesInternal<T>(
   const rendererFactory = lView[RENDERER_FACTORY];
 
+  const effectManager = lView[INJECTOR]?.get(EffectManager) ?? null;
+  setEffectManager(effectManager);
```
Needs to clean this up

dylhunn reviewed (approved):
reviewed-for: fw-core, public-api

dylhunn reviewed (approved):
reviewed-for: fw-core, public-api

atscott commented:
This PR was merged into the repository by commit ef149defb98aa9edba05359b27594594fe5b7617.
